cg.go

```diff
@@ -0,0 +1,162 @@
+package optimize
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+)
+
+const (
+	iterationRestartFactor = 6
+	angleRestartTreshold   = 0.9
+)
+
+type CGVariant interface {
+	Init(loc Location)
+	Beta(gradient, gradPrev, dirPrev []float64) float64
```

Comment:
Looking at the CG code, there should definitely be a comment on Beta specifying that zero is a special value.